### ISSUE TITLE
Transit read/write handlers

### ISSUE THREAD

**Issue by r0man** [gnarroway/hato#25]
Hello @gnarroway,

I can't use hato's conversion machinery with the Transit payloads in my application, because I defined custom Transit read/write handlers. It looks like hato does not support this at the moment. clj-http supports this with the `:transit-opts` configuration in the request map.

It would be nice if hato would support something like this as well.

WDYT?

Thanks for this library, r0man.

**Comment by gnarroway** [comment]
Thanks for raising this. I’ve never used those options so would be interested in whether they suit all use cases or whether you would make some modification to the way clj-http does it. If the former I might just use the same options.

**Comment by r0man** [comment]
@gnarroway I think the clj-http options would be sufficient. I used them in the past and did not need more. What's important though is that the options for the reader and writer are separate from each other. This is already the case, so I would just use the same options.

**Comment by gnarroway** [comment]
This is in 0.8.0-SNAPSHOT. Please try it out and see if it works for what you're doing.

Are you using `:as :auto` or one of the specific transit options? It looks like `auto` was not working, but it was already implemented for the others.

I've put some examples in the tests:

- [decoding](https://github.com/gnarroway/hato/blob/master/test/hato/middleware_test.clj#L224) (`:as :auto` requires server to provide content-type, or you can explicitly say `:transit+json` or `:transit+msgpack`)
- [encoding](https://github.com/gnarroway/hato/blob/master/test/hato/middleware_test.clj#L321) (requires `:form-params` and `:content-type`)

As an aside, to satisfy `:auto` you can pre-create a decoder and pass it in with the `:decoder` option if you need further customisation or don't want a decoder created [per request](https://github.com/gnarroway/hato/blob/master/src/hato/middleware.clj#L220)

**Comment by r0man** [comment]
Hi @gnarroway, 

thanks for looking into this! Unfortunately it does not work when doing a request to a server. The body of my response is nil. 

I'm doing the request as follows:

```
(hato.client/request
 {:scheme :http
  :server-name "localhost"
  :server-port 8001
  :accept "application/transit+msgpack"
  :as :transit+msgpack
  :content-type :transit+msgpack
  :form-params '[{(:burningswell/airports {:limit 2 :sort :name})
                  [:burningswell.list/title
                   {:burningswell.list/items
                    [:burningswell.airport/name]}]}]
  :request-method :post
  :throw-exceptions? false
  :uri "/api"})
```

This is the response:

```
{:request-time 32,
 :request
 {:throw-exceptions? false,
  :as :transit+msgpack,
  :headers
  {"content-type" "application/transit+msgpack",
   "accept" "application/transit+msgpack",
   "accept-encoding" "gzip, deflate"},
  :server-port 8001,
  :content-type "application/transit+msgpack",
  :http-request
  #object[jdk.internal.net.http.HttpRequestImpl 0x38272112 "http://localhost:8001/api POST"],
  :uri "/api",
  :server-name "localhost",
  :body
  [-111, -110, -90, 126, 35, 99, 109, 97, 112, -110, -110, -90, 126,
   35, 108, 105, 115, 116, -110, -73, 126, 58, 98, 117, 114, 110,
   105, 110, 103, 115, 119, 101, 108, 108, 47, 97, 105, 114, 112,
   111, 114, 116, 115, -126, -89, 126, 58, 108, 105, 109, 105, 116,
   2, -90, 126, 58, 115, 111, 114, 116, -90, 126, 58, 110, 97, 109,
   101, -110, -71, 126, 58, 98, 117, 114, 110, 105, 110, 103, 115,
   119, 101, 108, 108, 46, 108, 105, 115, 116, 47, 116, 105, 116,
   108, 101, -127, -71, 126, 58, 98, 117, 114, 110, 105, 110, 103,
   115, 119, 101, 108, 108, 46, 108, 105, 115, 116, 47, 105, 116,
   101, 109, 115, -111, -69, 126, 58, 98, 117, 114, 110, 105, 110,
   103, 115, 119, 101, 108, 108, 46, 97, 105, 114, 112, 111, 114,
   116, 47, 110, 97, 109, 101],
  :scheme :http,
  :request-method :post},
 :http-client
 #object[jdk.internal.net.http.HttpClientFacade 0x7c1c6320 "jdk.internal.net.http.HttpClientImpl@773f69f5(15)"],
 :headers
 {"x-download-options" "noopen",
  "x-permitted-cross-domain-policies" "none",
  "content-type" "application/transit+msgpack",
  "x-content-type-options" "nosniff",
  "x-frame-options" "DENY",
  "strict-transport-security" "max-age=31536000; includeSubdomains",
  "transfer-encoding" "chunked",
  "date" "Thu, 04 Feb 2021 16:16:15 GMT",
  "x-xss-protection" "1; mode=block",
  "content-security-policy"
  "object-src 'none'; script-src 'unsafe-inline' 'unsafe-eval' 'strict-dynamic' https: http:;"},
 :status 200,
 :content-type :application/transit+msgpack,
 :uri "http://localhost:8001/api",
 :content-type-params {},
 :version :http-1.1,
 :body nil}
```

I traced the issue down to these 2 lines here: 
- https://github.com/gnarroway/hato/blob/master/src/hato/middleware.clj#L47 (:as :transit+msgpack)
- https://github.com/gnarroway/hato/blob/master/src/hato/conversion.clj#L63 (:as :auto)

The call to `(.available in)` returns 0 and the body of the response is never parsed. I tried to remove this check and it seems to work fine with both modes. Do you think we can safely remove those checks?

Thanks Roman.

**Comment by gnarroway** [comment]
Thanks for tracing down the issue. Unfortunately I cannot reproduce this, with or without custom encoding/decoding, with or without auto (`.available` is always positive). What java are you running?

Here is a test you can add to `client_test.clj` that passes for me. Could you please see if passes for you and if so, could you try and make it fail?

```clojure
(deftest test-transit
  (testing "transit msgpack"
    (let [body {:a [1 2]}]
      (with-server (fn app [_]
                     {:status  200
                      :headers {"Content-Type" "application/transit+msgpack"}
                      :body    (let [output (ByteArrayOutputStream.)]
                                 (cognitect.transit/write (cognitect.transit/writer output :msgpack) body)
                                 (clojure.java.io/input-stream (.toByteArray output)))})

                   (let [r (get "http://localhost:1234" {:as :auto})]
                     (is (= body (:body r))))))))
```

That said, the javadoc of InputStream says `available` will always return 0 (for the abstract class) so it could be some artefact of the type hint. But first I would like to be able to reproduce your error.

**Comment by gnarroway** [comment]
Whether it is safe to remove the offending line, there is discussion about when it was added to both aleph and cli-http https://github.com/clj-commons/aleph/pull/331

**Comment by r0man** [comment]
@gnarroway Thanks again for looking into this. I just run the test you posted above on my machine, and it looks like it is a JDK version issue. I was using JDK 11 when I got `nil` as a response. The test you posted above fails like this:

```
lein test :only hato.client-test/test-transit

FAIL in (test-transit) (client_test.clj:309)
transit msgpack
expected: (= body (:body r))
  actual: (not (= {:a [1 2]} nil))
```
On JDK 15 the test passes, and I do get the expected response. I also had GraalVM version of the JDK 11 and it failed there as well.

I'm running this on:

`Linux thinkpad 5.10.13-arch1-1 #1 SMP PREEMPT Wed, 03 Feb 2021 23:44:07 +0000 x86_64 GNU/Linux`

### Working JDKs

openjdk version "15.0.2" 2021-01-19
OpenJDK Runtime Environment (build 15.0.2+7)
OpenJDK 64-Bit Server VM (build 15.0.2+7, mixed mode)

### Failing JDKs

openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment (build 11.0.10+9)
OpenJDK 64-Bit Server VM (build 11.0.10+9, mixed mode)

openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment GraalVM CE 21.0.0 (build 11.0.10+8-jvmci-21.0-b06)
OpenJDK 64-Bit Server VM GraalVM CE 21.0.0 (build 11.0.10+8-jvmci-21.0-b06, mixed mode, sharing)

**Comment by gnarroway** [comment]
Thanks for taking the time again.

- I have checked in a change that works on both JDK versions and handles the empty stream using a `try`. 
- I've also raised #27 to more proactively catch such errors in future.

Let me know if the snapshot build works for you now.

**Comment by r0man** [comment]
@gnarroway It works now on both JDK versions. Thank you!

**Comment by gnarroway** [comment]
released in 0.8.0!

**Comment by r0man** [comment]
Perfect. Thanks again!
